trunk/source/CACSL2BoogieTranslator/src/de/uni_freiburg/informatik/ultimate/plugins/generator/cacsl2boogietranslator/Boogie2ACSL.java

```diff
@@ -0,0 +1,657 @@
+
+	private static String bitvecToString(final BitvecLiteral lit) {
+		final String binStr = new BigInteger(lit.getValue()).toString(2);
+		assert binStr.length() <= lit.getLength() : "Binary string cannot be longer than bitvector literal length";
+		final int missingZeros = lit.getLength() - binStr.length();
+		if (missingZeros > 0) {
+			final String formatStr = "%" + lit.getLength() + "s";
+			return String.format(formatStr, binStr).replace(' ', '0');
+		}
+		return binStr;
+	}
+
+	// TODO: Currently we don't care about types here, since function applications should only occur in bitvector
+	// setting, where we should not need any casts.
+	private BacktranslatedExpression translateFunctionApplication(
+			final de.uni_freiburg.informatik.ultimate.boogie.ast.FunctionApplication fun, final ILocation context) {
+		final BacktranslatedExpression[] translatedArguments = new BacktranslatedExpression[fun.getArguments().length];
+		for (int i = 0; i < fun.getArguments().length; i++) {
+			translatedArguments[i] = translateExpression(fun.getArguments()[i], context);
+			if (translatedArguments[i] == null) {
+				return null;
+			}
+		}
+		final String function = fun.getIdentifier();
+		final Pair<String, Integer> reversed = SFO.reverseBoogieFunctionName(function);
+		if (reversed == null) {
+			mReporter.accept("Cannot identify Boogie2SMT function " + function);
+			return null;
+		}
+		final Integer bitSize = reversed.getSecond();
+
+		switch (reversed.getFirst()) {
+		case "sign_extend":
+		case "zero_extend":
+			// TODO: This might be problematic for signed types!
+			return translatedArguments[0];
+		case "fp":
+			if (Arrays.stream(fun.getArguments()).allMatch(BitvecLiteral.class::isInstance)) {
+				// this function is used to construct floating points
+				return constructFloat((BitvecLiteral) fun.getArguments()[0], (BitvecLiteral) fun.getArguments()[1],
+						(BitvecLiteral) fun.getArguments()[2]);
+			}
+			mReporter.accept("fp can only be backtranslated, if the arguments are literals: " + fun);
+			return null;
+		case "NaN":
+			return new BacktranslatedExpression(new RealLiteral(String.valueOf(Float.NaN)));
+		case "bvadd":
+			final Expression addition = new BinaryExpression(Operator.ARITHPLUS, translatedArguments[0].getExpression(),
+					translatedArguments[1].getExpression());
+			return new BacktranslatedExpression(new BinaryExpression(Operator.ARITHMOD, addition,
+					new IntegerLiteral(String.valueOf(1L << bitSize))));
+		case "bvmul":
+			return new BacktranslatedExpression(new BinaryExpression(Operator.ARITHMUL,
+					translatedArguments[0].getExpression(), translatedArguments[1].getExpression()));
+		case "bvsub":
+			final Expression sub = new BinaryExpression(Operator.ARITHMINUS, translatedArguments[0].getExpression(),
+					translatedArguments[1].getExpression());
+			return new BacktranslatedExpression(
+					new BinaryExpression(Operator.ARITHMOD, sub, new IntegerLiteral(String.valueOf(1L << bitSize))));
+		case "bvand":
```

Comment:
Yes, we could just set the operator here. I am not quite sure, if this is much better. Maybe some of the bitvector-operators might require some special handling (like modulo), so this might also change in the future. I just adapted the code from `CACSL2BoogieBacktranslator`.